src/package/packageConvert.ts

```diff
@@ -136,29 +145,72 @@ export async function convertPackage(
  * @private
  */
 export async function createPackageVersionCreateRequest(
-  context: { installationkey?: string; buildinstance?: string },
-  packageId: string
+  context: { installationkey?: string; definitionfile?: string; buildinstance?: string },
+  packageId: string,
+  apiVersion: string
 ): Promise<PackagingSObjects.Package2VersionCreateRequest> {
   const uniqueId = uniqid({ template: `${packageId}-%s` });
   const packageVersTmpRoot = path.join(os.tmpdir(), uniqueId);
+  const packageVersMetadataFolder = path.join(packageVersTmpRoot, 'md-files');
   const packageVersBlobDirectory = path.join(packageVersTmpRoot, 'package-version-info');
+  const settingsZipFile = path.join(packageVersBlobDirectory, 'settings.zip');
+  const metadataZipFile = path.join(packageVersBlobDirectory, 'package.zip');
   const packageVersBlobZipFile = path.join(packageVersTmpRoot, consts.PACKAGE_VERSION_INFO_FILE_ZIP);
 
   const packageDescriptorJson = {
     id: packageId,
   };
 
+  const settingsGenerator = new SettingsGenerator({ asDirectory: true });
+  const definitionFile = context.definitionfile;
+
+  try {
+    if (definitionFile) {
+      const definitionFilePayload = await fs.promises.readFile(definitionFile, 'utf8');
+      const definitionFileJson = JSON.parse(definitionFilePayload) as ScratchOrgInfo;
+
+      // Load any settings from the definition
+      await settingsGenerator.extract(definitionFileJson);
+      if (settingsGenerator.hasSettings() && definitionFileJson.orgPreferences) {
+        // this is not allowed, exit with an error
+        throw messages.createError('signupDuplicateSettingsSpecified');
+      }
+
+      ['country', 'edition', 'language', 'features', 'orgPreferences', 'snapshot', 'release', 'sourceOrg'].forEach(
+        (prop) => {
+          const propValue = definitionFileJson[prop];
+          if (propValue) {
+            packageDescriptorJson[prop] = propValue;
+          }
+        }
+      );
+    }
+  } catch (err) {
+    throw messages.createError('errorReadingDefintionFile', [err as string]);
+  }
+
   await fs.promises.mkdir(packageVersTmpRoot, { recursive: true });
   await fs.promises.mkdir(packageVersBlobDirectory, { recursive: true });
+  await fs.promises.mkdir(packageVersMetadataFolder, { recursive: true });
 
+  await settingsGenerator.createDeploy();
+  await settingsGenerator.createDeployPackageContents(apiVersion);
+  await srcDevUtil.zipDir(
+    `${settingsGenerator.getDestinationPath()}${path.sep}${settingsGenerator.getShapeDirName()}`,
+    settingsZipFile
+  );
+
+  const shapeDirectory = `${settingsGenerator.getDestinationPath()}${path.sep}${settingsGenerator.getShapeDirName()}`;
+  const currentPackageXml = await fs.promises.readFile(path.join(shapeDirectory, 'package.xml'), 'utf8');
```

Comment:
@smotwani11 is package.xml file guaranteed to be at the location produced by the path.join? If not, the read should be in a try/catch.

Reply:
yes package.xml will be generated by the settings generator. The tests which don't provide the definition file option cover that scenario. The package.xml will be present in that location.